rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,316 @@
+message YubikeyPIVAttestationRequest {
+  bytes slot_cert = 1;
+  bytes attestation_cert = 2;
+}
+
+message AttestHardwarePrivateKeyResponse { }
+```
+
+In addition to verifying the certificate chain for a user's hardware private key to a trusted hardware key manufacturer, the resulting attestation object will provide information about the private key, including;
+ - Device information, including serial number, model, version 
+ - Configured Touch (And PIN) Policies if any
+
+This attestation data will be check against server settings. If the attestation is valid, it will be stored in the backend in `/hardware_key_attestations/<public_key_der>`.
+
+```go
+type AttestationData struct {
+  // PublicKeyDER is the public key of the hardware private key attested in PKIX, ASN.1 DER form.
+	PublicKeyDER []byte `json:"public_key"`
+  // Type is the type of hardware key this attestation is for.
+	Type HardwareKeyType `json:"type"`
+	// TouchRequired specifies whether touch is required to access the hardware private key.
+	TouchRequired bool `json:"touch_required"`
+  // json blob containing the resulting attestation object. 
+  //
+  // For yubikey, this will look like https://pkg.go.dev/github.com/go-piv/piv-go@v1.9.0/piv#Attestation
+  // In the future, we can check the attesation object for additional information, like PIN policy,
+  // model type, serial number, etc.
+  AttestationObject []byte `json:"attestation"`
+}
+```
+
+### Certificate private key policy extension
+
+When the Auth Server receives a new certificate request, it will check the cluster auth preference and the user's role options to get their required private key policy. If their policy is `hardware_key`, then the Auth Server will look for a valid attestation in the backend associated with the given public key. If their policy is `hardware_key_touch`, then the Auth Server will also check that the attestation has `TouchRequired: true`.
+
+If the attestation checks are successful, then the Auth Server will sign the TLS and SSH certificates with the new `private_key_policy` TLS extra subject name and Extension respectively. The Auth Server's Authorizer, which is used for all API requests, will check certificates for this extension before authorizing requests.
+
+### Client changes
+
+Teleport clients will need the ability to connect to a user's hardware key, generate/retrieve private keys, and use those keys for cryptographical operations. 
+
+#### Private key policy discovery
+
+Teleport clients should be able to automatically determine if a user requires a hardware private key for login to avoid additional UX concerns. For this, we will introduce the `GetPrivateKeyPolicy` rpc.
+
+```proto
+service AuthService {
+  rpc GetPrivateKeyPolicy(GetPrivateKeyPolicyRequest) returns (GetPrivateKeyPolicyResponse);
+}
+
+// The user will be pulled from the request certificate, so request message can be empty for now.
+message GetPrivateKeyPolicyRequest {}
```

Comment:
If we are tying policies to session MFA checks, doesn't that mean that the reply could change based on the resource being accessed?

Or, is it an all-or-nothing check, meaning that if any of my roles have require_session_mfa=hardware_key, then suddenly that constrains _all_ my access?

It looks like this is one of the issues we get by overloading the require_session_mfa check, it doesn't exactly fit.

Reply:
It will have to constrain all of your access, since the certificates are shared across your entire login session, rather than the single use certificates used for a single session request. As mentioned in the RFD, the implementation is completely separate from the concept of per-session mfa checks, so we'll have to clear it up in the documentation as best as we can.

Reply:
Ack. Yep, good docs will surely help here.